gossipd/gossip_store.c

```diff
@@ -24,6 +31,15 @@ void gossip_store_append(struct gossip_store *gs, const u8 *msg)
 	u16 msglen = tal_len(msg);
 	beint16_t belen = CPU_TO_BE16(msglen);
 
+	/* FIXME: this method of detecting replayed messages is best effort
```

Comment:
This seems hackish to me.  Perhaps we could do instead:

1.  Split `handle_gossip_message` into `handle_gossip_message_and_store` and `handle_gossip_message`.
    - `handle_gossip_message_and_store` calls `handle_gossip_message`.
    - `handle_gossip_message` adds a new output argument, `bool *new_message`, which indicates whether the given gossip message is new/fresh.
    - In `handle_gossip_message_and_store`, only store the gossip message if `handle_gossip_message` did not return an error and returned a false `new_message` flag.  This is the only point at which gossip messages are stored.
2. peer and master message handlers call `handle_gossip_message_and_store` instead.
3. The individual `handle_cannouncement` etc do not store the message.  Instead they also have a `bool *new_message` output argument that they clear if the message has been received before or set if the message is not encountered before.
4. The gossip store replaying uses `handle_gossip_message`.

Reply:
I did that at first, with an additional `store` argument, but due to the way we weave the handling through async calls, we'd have to drag that value through everything as well. In the end I had to add the `store` value in 20+ places, which seemed absolute overkill. The reason we need to weave it through is that only at the end of the verification do we actually know whether this is a good message that should be stored or not.

The new output argument you mention does not work exactly because we may need to call out to `masterd` to verify a `txout` or we may have been queuing behind a `channel_announcement` being verified. This just seemed like the fix with the minimal impact.